compiler/luci/pass/src/FoldSparseToDensePass.cpp

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2021 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/FoldSparseToDensePass.h"
+#include "CircleOptimizerUtils.h"
+
+#include <luci/IR/CircleNodes.h>
+
+namespace
+{
+
+/**
+ * Fold to const if
+ *
+ * 1. indices has 0-sized static shape such as [0]
+ *    (i.e., output is filled with default value)
+ * 2. default_value: const scalar
+ * 3. output_shape: const
+ *
+ * TODO: Support more general patterns
+ **/
+template <loco::DataType IndexT, loco::DataType ValueT>
+bool fold_sparse_to_dense(luci::CircleSparseToDense *stod)
+{
+  const auto indices = loco::must_cast<luci::CircleNode *>(stod->indices());
+  const auto default_value = loco::must_cast<luci::CircleConst *>(stod->default_value());
+  const auto output_shape = loco::must_cast<luci::CircleConst *>(stod->output_shape());
+
+  int32_t indices_size = 1;
+  for (uint32_t i = 0; i < indices->rank(); i++)
+  {
+    if (!indices->dim(i).known())
+      return false;
+    indices_size *= indices->dim(i).value();
+  }
+
+  if (indices_size != 0)
+    return false;
```

Comment:
I think that even some dimensions are unknown, some cases are possible to be folded.
For example, if shape of indices is [?, 0, ?, 1], it is definitely zero sized shape.
So how about following codes?

```suggestion
  bool has_zero = false;
  for (uint32_t i = 0; i < indices->rank(); i++)
  {
    if (indices->dim(i).known() && indices->dim(i).value() == 0)
      has_zero = true;
  }

  if (!has_zero)
    return false;
```

Reply:
Good idea :) I've updated the code. Thanks.